fs/zbd_zenfs.cc

```diff
@@ -513,22 +529,41 @@ Zone *ZonedBlockDevice::AllocateZone(Env::WriteLifeTimeHint file_lifetime) {
       if (finish_victim == nullptr) {
         finish_victim = z;
       } else if (finish_victim->capacity_ > z->capacity_) {
+        assert(finish_victim->UnsetBusy());
         finish_victim = z;
+      } else {
+        assert(z->UnsetBusy());
       }
+    } else {
+      assert(z->UnsetBusy());
     }
   }
 
+  // Holding finish_victim if != nullptr
+
   /* Try to fill an already open zone(with the best life time diff) */
   for (const auto z : io_zones) {
-    if ((!z->open_for_write_) && (z->used_capacity_ > 0) && !z->IsFull()) {
-      unsigned int diff = GetLifeTimeDiff(z->lifetime_, file_lifetime);
-      if (diff <= best_diff) {
-        allocated_zone = z;
-        best_diff = diff;
+    if (z->SetBusy()) {
+      if ((z->used_capacity_ > 0) && !z->IsFull()) {
+        unsigned int diff = GetLifeTimeDiff(z->lifetime_, file_lifetime);
+        if (diff <= best_diff) {
+          if (allocated_zone != nullptr) {
+            assert(allocated_zone->UnsetBusy());
+          }
+          allocated_zone = z;
+          best_diff = diff;
+        } else {
+          assert(z->UnsetBusy());
```

Comment:
Danger danger! everything inside assert() will be removed when built DEBUG_LEVEL==0 , not what we want, right?

Reply:
Good catch! Will change that. This also reveals that I only ran the test suite with debug mode enabled :confounded:

Reply:
This PR is addressing the problem: https://github.com/metaspace/zenfs/pull/1